### PR TITLE
driver: RIOT port for the Freescale kw2x radio

### PR DESCRIPTION
**Pull request by jremmert-phytec-iot**
This PR adds support for Freescale KW2x radio. The driver uses the new netdev interface exclusively, the old transceiver interface is not supported.

For the use with: MKW24D512, MKW22D512, MKW21D512, and
MKW21D256

This PR depends on #2059, ~~#2265~~ and #2188.

### CODE REVIEW

File: drivers/include/kw2xrf.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+}
+kw2xrf_packet_t;
+
+/**
+ * Netdev representation of this driver.
+ */
+extern netdev_t kw2xrf_netdev;
+
+
+
```
Comment (thomaseichinger) [inline drivers/include/kw2xrf.h]: Too many newlines.

File: drivers/include/kw2xrf.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+/** Utility macro: get KW2XRF's status byte. */
+#define kw2xrf_status_byte() kw2xrf_strobe(NOBYTE)
```
Comment (thomaseichinger) [inline drivers/include/kw2xrf.h]: Please change macro name to upper case.

File: drivers/kw2xrf/include/kw2xrf_reg.h [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+#define MKW2XDM_PWR_MODES_XTALEN                    (1 << 4)
+#define MKW2XDM_PWR_MODES_AUTODOZE                  (1 << 1)
+#define MKW2XDM_PWR_MODES_PMC_MODE                  (1 << 0)
+
+enum mkw2xdrf_iregister {
+    MKW2XDMI_PART_ID = 0x00,		 /* Part Identification */
+    MKW2XDMI_XTAL_TRIM = 0x01,		 /* XTAL 32 MHz Trim */
+    MKW2XDMI_MACPANID0_LSB = 0x03,	 /* MAC PAN ID for PAN0 */
+    MKW2XDMI_MACPANID0_MSB = 0x04,	 /* MAC PAN ID for PAN0 */
+    MKW2XDMI_MACSHORTADDRS0_LSB = 0x05,	 /* MAC Short Address for PAN0 */
```
Comment (thomaseichinger) [inline drivers/kw2xrf/include/kw2xrf_reg.h]: Here and all following, please align comments.

File: drivers/kw2xrf/include/kw2xrf_reg.h [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+#define MKW2XDMI_PART_ID_MANUF_ID_MASK			0x60u
+#define MKW2XDMI_PART_ID_VERSION_MASK			0x1Cu
+#define MKW2XDMI_PART_ID_MASK_SET_MASK			0x07u
+
+#define MKW2XDMI_RX_FRAME_FILTER_FRM_VER_MASK		0xC0u
+#define MKW2XDMI_RX_FRAME_FILTER_FRM_VER_SHIFT		6
+#define MKW2XDMI_RX_FRAME_FILTER_FRM_VER(x)             (((uint8_t)(((uint8_t)(x))<<MKW2XDMI_RX_FRAME_FILTER_FRM_VER_SHIFT))&MKW2XDMI_RX_FRAME_FILTER_FRM_VER_MASK)
+#define MKW2XDMI_RX_FRAME_FILTER_ACTIVE_PROMISCUOUS	(1 << 5)
+#define MKW2XDMI_RX_FRAME_FILTER_NS_FT			(1 << 4)
+#define MKW2XDMI_RX_FRAME_FILTER_CMD_FT			(1 << 3)
```
Comment (thomaseichinger) [inline drivers/kw2xrf/include/kw2xrf_reg.h]: Please replace tabs with spaces.

File: drivers/kw2xrf/include/kw2xrf_reg.h [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+#define MKW2XDMI_GPIO_DS8				(1 << 7)
+#define MKW2XDMI_GPIO_DS7				(1 << 6)
+#define MKW2XDMI_GPIO_DS6				(1 << 5)
+#define MKW2XDMI_GPIO_DS5				(1 << 4)
+#define MKW2XDMI_GPIO_DS4				(1 << 3)
+#define MKW2XDMI_GPIO_DS3				(1 << 2)
+#define MKW2XDMI_GPIO_DS2				(1 << 1)
+#define MKW2XDMI_GPIO_DS1				(1 << 0)
+
+#define MKW2XDMI_ANT_PAD_CTRL_ANTX_POL3			(1 << 7)
```
Comment (thomaseichinger) [inline drivers/kw2xrf/include/kw2xrf_reg.h]: s/\t/    /

File: drivers/kw2xrf/kw2xrf.c [hunk 6]
```diff
@@ -0,0 +1,855 @@
...
+    -23, -22, -20, -19,
+    -17, -16, -14, -13,
+    -11, -10, -8, -7,
+    -5, -4, -2, -1,
+    1, 2, 4, 5,
+    7
+};
+
+/* implementation of driver's functions */
+
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]: newline

File: drivers/kw2xrf/kw2xrf.c [hunk 7]
```diff
@@ -0,0 +1,855 @@
...
+{
+    kw2xrf_write_dreg(MKW2XDM_IRQSTS1, 0xff);
+    kw2xrf_write_dreg(MKW2XDM_IRQSTS2, MKW2XDM_IRQSTS2_PB_ERR_IRQ
+                      | MKW2XDM_IRQSTS2_WAKE_IRQ);
+    kw2xrf_write_dreg(MKW2XDM_IRQSTS3, 0x0f);
+    DEBUG("IRQSTS1: %x\n", kw2xrf_read_dreg(MKW2XDM_IRQSTS1));
+    DEBUG("IRQSTS2: %x\n", kw2xrf_read_dreg(MKW2XDM_IRQSTS2));
+    DEBUG("IRQSTS3: %x\n", kw2xrf_read_dreg(MKW2XDM_IRQSTS3));
+    uint8_t reg = kw2xrf_read_dreg(MKW2XDM_PHY_CTRL2);
+    //reg &= ~(MKW2XDM_PHY_CTRL2_RXMSK | MKW2XDM_PHY_CTRL2_SEQMSK);
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]: Please remove if not needed.

File: drivers/kw2xrf/kw2xrf.c [hunk 8]
```diff
@@ -0,0 +1,855 @@
...
+    -5, -4, -2, -1,
+    1, 2, 4, 5,
+    7
+};
+
+/* implementation of driver's functions */
+
+void kw2xrf_init_interrupts(void)
+{
+    kw2xrf_write_dreg(MKW2XDM_IRQSTS1, 0xff);
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]: Please add some comments describing what's done here.

File: drivers/kw2xrf/kw2xrf.c [hunk 9]
```diff
@@ -0,0 +1,855 @@
...
+
+int kw2xrf_initialize(netdev_t *dev)
+{
+    if (dev == NULL || dev->driver != ((netdev_driver_t *)(&kw2xrf_driver))) {
+        return -ENODEV;
+    }
+
+    dev->type = NETDEV_TYPE_802154;
+    dev->more = NULL;
+    //dev->event_handler = thread_getpid();
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]: Remove if not needed.

File: drivers/kw2xrf/kw2xrf.c [hunk 10]
```diff
@@ -0,0 +1,855 @@
...
+    kw2xrf_set_tx_power(0);
+#ifdef MODULE_CONFIG
+    radio_pan = sysconfig.radio_pan_id;
+#else
+    radio_pan = 0x0001;
+#endif
+    kw2xrf_set_pan(radio_pan);
+    radio_channel = MKW2XDRF_DEFAULT_CHANNR;
+    kw2xrf_set_channel(radio_channel);
+    DEBUG("MKW2XDRF initialized and set to channel %i and pan 0x1111\n", MKW2XDRF_DEFAULT_CHANNR);
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]: `0x1111` seems wrong.

File: drivers/kw2xrf/include/kw2xrf_internal.h [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+#include "periph_conf.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Initialize the KW2XRF transceiver.
+ *
+ * @param[in] dev The netdev network device.
```
Comment (thomaseichinger) [inline drivers/kw2xrf/include/kw2xrf_internal.h]: missing `@return`

File: drivers/kw2xrf/kw2xrf.c [hunk 12]
```diff
@@ -0,0 +1,855 @@
...
+     *
+     * Equation for PLL frequency, MKW2xD Reference Manual, p.255 :
+     * F = ((PLL_INT0 + 64) + (PLL_FRAC0/65536))32MHz
+     *
+     */
+    ch -= 11;
+    kw2xrf_write_dreg(MKW2XDM_PLL_INT0, MKW2XDM_PLL_INT0_VAL(pll_int_lt[ch]));
+    kw2xrf_write_dreg(MKW2XDM_PLL_FRAC0_LSB, (uint8_t)pll_frac_lt[ch]);
+    kw2xrf_write_dreg(MKW2XDM_PLL_FRAC0_MSB, (uint8_t)(pll_frac_lt[ch] >> 8));
+    return ((unsigned int) ch);
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]: Better return `radio_channel` here.

File: drivers/kw2xrf/kw2xrf.c [hunk 13]
```diff
@@ -0,0 +1,855 @@
...
+#else
+    return radio_channel;
+#endif
+}
+
+void kw2xrf_set_monitor(int mode)
+{
+    uint8_t tmp;
+    tmp = kw2xrf_read_dreg(MKW2XDM_PHY_CTRL4);
+    tmp |= MKW2XDM_PHY_CTRL4_PROMISCUOUS;
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]:
```C
if (mode) {
    tmp |= MKW2XDM_PHY_CTRL4_PROMISCUOUS;
} else {
    tmp &= ~MKW2XDM_PHY_CTRL4_PROMISCUOUS;
}
```

File: drivers/kw2xrf/kw2xrf.c [hunk 14]
```diff
@@ -0,0 +1,855 @@
...
+{
+    uint8_t tmp;
+    tmp = kw2xrf_read_dreg(MKW2XDM_PHY_CTRL4);
+    return ((tmp & MKW2XDM_PHY_CTRL4_PROMISCUOUS) != 0);
+}
+
+
+int kw2xrf_set_tx_power(int pow)
+{
+    if (pow > MKW2XDRF_OUTPUT_POWER_MAX) {
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]: The function description states values -25 to 0 while the defines range from -35 to 8?

File: drivers/kw2xrf/kw2xrf.c [hunk 15]
```diff
@@ -0,0 +1,855 @@
...
+        case NETDEV_OPT_PROTO:
+            if (*value_len < sizeof(netdev_proto_t)) {
+                return -EOVERFLOW;
+            }
+
+            if (*value_len > sizeof(netdev_proto_t)) {
+                *value_len = sizeof(netdev_proto_t);
+            }
+
+            *((netdev_type_t *)value) = NETDEV_PROTO_802154;
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]: s/netdev_type_t/netdev_proto_t/

File: drivers/kw2xrf/kw2xrf_rx.c [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+    pkt_len = kw2xrf_read_dreg(MKW2XDM_RX_FRM_LEN);
+
+    /* read PSDU */
+    kw2xrf_read_fifo(buf, pkt_len + 1);
+    pkt_lqi = buf[pkt_len];
+    pkt_rssi = 0;
+
+    /* low-level reception mechanism (for MAC layer, among others) */
+    if ((recv_func.type == MKW2XDRF_CB_TYPE_RAW) && (recv_func.cb.raw != NULL)) {
+        recv_func.cb.raw(NULL, buf, pkt_len - 2, pkt_rssi, pkt_lqi, crc_ok);
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_rx.c]: Does the received packet not receive a CRC flag? I don't see `crc_ok` changing since L:121.

File: drivers/kw2xrf/kw2xrf_rx.c [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+                          &kw2xrf_rx_buffer[rx_buffer_next].frame,
+                          pkt_len);
+
+    /* follow-up to transceiver module if adequate */
+    if (kw2xrf_rx_buffer[rx_buffer_next].frame.fcf.frame_type != 2) {
+#ifdef DEBUG
+        ieee802154_frame_print_fcf_frame(&kw2xrf_rx_buffer[rx_buffer_next].frame);
+#endif
+
+#ifdef MODULE_TRANSCEIVER
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_rx.c]: Remove this block if you don't plan to support the transceiver module.

File: drivers/kw2xrf/kw2xrf_spi.c [hunk 18]
```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (C) 2015555555555555555555555555555555555555555555555555555555 PHYTEC Messtechnik GmbH
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_spi.c]: You sent this from the future?

File: drivers/kw2xrf/kw2xrf_spi.c [hunk 19]
```diff
@@ -0,0 +1,277 @@
...
+#include "cpu-conf.h"
+#include "periph/gpio.h"
+#include "irq.h"
+
+/* SPI1 connected to the modem (ieee802.15.4 radio) */
+#define KW2XDRF_SPI                   SPI1
+#define KW2XDRF_SPI_CLKEN()           (SIM->SCGC6 |= (SIM_SCGC6_SPI1_MASK))
+#define KW2XDRF_SPI_CLKDIS()          (SIM->SCGC6 &= ~(SIM_SCGC6_SPI1_MASK))
+#define KW2XDRF_SPI_IRQ               SPI1_IRQn
+#define KW2XDRF_SPI_IRQ_HANDLER       isr_spi1
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_spi.c]: Please move these to `periph_conf.h`.

File: drivers/kw2xrf/kw2xrf_spi.c [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+
+void kw2xrf_spi_init(void)
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_spi.c]: Please move SPI initialisation and transfer functions to `periph/spi.c`.

File: drivers/kw2xrf/kw2xrf_tx.c [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+#include "kw2xrf.h"
+#include "kw2xrf_reg.h"
+#include "kw2xrf_spi.h"
+#include "kw2xrf_internal.h"
+#include "ieee802154_frame.h"
+
+#include "irq.h"
+#include "hwtimer.h"
+
+#define ENABLE_DEBUG    (1)
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_tx.c]: Please turn DEBUG off.

File: drivers/kw2xrf/kw2xrf.c [hunk 22]
```diff
@@ -0,0 +1,855 @@
...
+#include "kw2xrf_reg.h"
+#include "kw2xrf_internal.h"
+#include "hwtimer.h"
+#include "kernel_types.h"
+#include "netdev/802154.h"
+#include "netdev/base.h"
+#include "thread.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG    (1)
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]: Please turn DEBUG off.

File: drivers/kw2xrf/kw2xrf_tx.c [hunk 23]
```diff
@@ -0,0 +1,321 @@
...
+
+    if (size > MKW2XDRF_MAX_PKT_LENGTH) {
+        return NETDEV_802154_TX_STATUS_PACKET_TOO_LONG;
+    }
+
+    pkt[0] = size;
+
+    /* generate pkt */
+    for (uint8_t i = 0; i < len; i++) {
+        hdr[i + idx] = ((uint8_t *)buf)[i];
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_tx.c]: This looks weird, could use `pkt` here?

File: drivers/kw2xrf/kw2xrf_tx.c [hunk 24]
```diff
@@ -0,0 +1,321 @@
...
+
+    /* XXX TEST XXX */
+    /*
+    kw2xrf_write_fifo(test, 9);
+    kw2xrf_read_fifo(pkt, packet->length - 1);
+    for (uint8_t i = 0; i < packet->length-1; i++) {
+    	DEBUG("tx: fifo: %d: %x\n", i, pkt[i]);
+    }
+    */
+    /* XXX TEST end XXX */
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_tx.c]: Remove unused code.

File: drivers/kw2xrf/kw2xrf_tx.c [hunk 25]
```diff
@@ -0,0 +1,321 @@
...
+#define ACK_LENGTH  5
+
+netdev_802154_tx_status_t kw2xrf_transmit_tx_buf(netdev_t *dev)
+{
+    /* check if channel clear */
+    if (!kw2xrf_channel_clear(dev)) {
+        return NETDEV_802154_TX_STATUS_MEDIUM_BUSY;
+    }
+
+    // while(13){
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_tx.c]: I see dead code.

Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_tx.c]: Please clean up this whole block.

File: drivers/kw2xrf/kw2xrf_tx.c [hunk 26]
```diff
@@ -0,0 +1,321 @@
...
+
+    kw2xrf_write_dreg(MKW2XDM_IRQSTS1, MKW2XDM_IRQSTS1_TXIRQ);
+
+    /*TODO check seqnum */
+    /* return RADIO_TX_OK; */
+    return NETDEV_802154_TX_STATUS_NOACK;
+}
+
+/* TODO: add global modem cfg_struct for addresses and settings */
+int16_t kw2xrf_send(kw2xrf_packet_t *packet)
```
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_tx.c]: This function is only needed for `transceiver.h`.

**Comment by thomaseichinger** [comment]
@jremmert-phytec-iot @jfischer-phytec-iot  Awesome work! Please address the comments above. 
@haukepetersen Do you have the board to test this?

File: drivers/kw2xrf/kw2xrf_spi.c (hunk 20 above)
Comment (jfischer-phytec-iot) [inline drivers/kw2xrf/kw2xrf_spi.c]: @thomaseichinger I do not want, even if we should avoid the redundancy of the code,
as long as it is not possible to use HW CS and spi interfaces are not cleanly separated.

The modem is inside the SiP an is internally connected via the SPI1 to the CPU. It is not possible to use the SPI1 for what else.

**Comment by jfischer-phytec-iot** [comment]
@thomaseichinger If more Board are needed, please send me a mail, we have a few in stock currently.

File: drivers/kw2xrf/kw2xrf_spi.c (hunk 20 above)
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_spi.c]: Well I'd guess that's the crux with hardware abstraction, it's hard to match all platforms' specialties. Some remarks

1. I don't see any logic in here which can not be used with the current SPI interface and mapped by macros. (If the Freescale SPI can be used without external CS I'm fine with it. Using GPIOs is no requirement)
2. If there is an interface for some functionality I think we should use it to minimise clutter. (`periph/spi.h` is always open for discussion, and I'd be happy if we find extensions to it while keeping it simple.)
3. In this current version the driver is not portable to other base MCUs and SPI specific code in different locations.
4. In my humble opinion I don't see advantages for HW CS if not used with DMA, especially in combination with turned off interrupts while busy waiting for transfers to finish.

To sum up, I'd very welcome if you could reconsider your decisions.

File: drivers/kw2xrf/kw2xrf.c (hunk 14 above)
Comment (jremmert-phytec-iot) [inline drivers/kw2xrf/kw2xrf.c]: @thomaseichinger Thank you for reviewing! But I´ve there is a question at this point. 
The KW2x-rf radio is capable of power-values from -35 to +8dbm. These power-values are mapped to the internal radio-register (Modem_PA_PWR). The useable range of this register is defined from 3-31 in the reference manual.
What do you mean when you said the function description states values of -25 to 0? Is that an additional interface definition? Is it correct that the parameter of the *set_option->NETDEV_OPT_TX_POWER function represents the physical output power in dBm?

Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf.c]: Sorry, I wasn't clear about this. I was referencing the parameter description at [kw2xrf_internal.h:L156](https://github.com/RIOT-OS/RIOT/pull/2329/files#diff-05df4e9a72f4a10d9e34af8824d2a838R156)

File: drivers/kw2xrf/kw2xrf_rx.c (hunk 16 above)
Comment (jremmert-phytec-iot) [inline drivers/kw2xrf/kw2xrf_rx.c]: The modem automatically checks the CRC in hardware. An interrupt is only triggered in case of a valid crc. Therefore the crc_ok is always true. But as it´s not obviously I commented this context and deleted the crc_ok variable as it is not really used.

File: drivers/kw2xrf/kw2xrf.c (hunk 14 above)
Comment (jremmert-phytec-iot) [inline drivers/kw2xrf/kw2xrf.c]: Oh I forgot to adapt the interface description. I will adapt the description to -35 to 8 (dBm).

File: drivers/kw2xrf/kw2xrf_spi.c (hunk 20 above)
Comment (jremmert-phytec-iot) [inline drivers/kw2xrf/kw2xrf_spi.c]: For the moment we decided to continue using the internal SPI-implementation in the driver implementation, because the complexity is far lower with the current implementation. *But we think about extending the SPI-driver in kinetis_common and use that for the radio-driver implementation.* We have some concerns about using the RIOT-low-level SPI. 

- The spi interface requires CS, but we don`t want to use it as GPIO because the SiP hardware function blocks provides a direct CS. The timings might also depend on the GPIO-number in case of GPIO-CS. The use of the hardware-integrated CS makes the communication more energy efficient.
- It is not possible to use the radio with any other base MCU than the integrated M4-Kinetis MCU, as it is an integrated System im Package. Other potential Freescale stand-alone radio transceivers have different register setup, so anyway they will not be compatible to this driver.
- We would like to use DMA. Not right now, but there will be an improvement that addresses the DMA capabilities.
- The "user" SPI-Interface is different from the radio SPI-interface. (SPI0 (user SPI) vs. SPI1 (radio-spi)).
- In kinetis_common there is an seperate implementation of the Low-level RIOT-spi interface, that can be used for user purposes.

File: drivers/kw2xrf/kw2xrf_tx.c (hunk 23 above)
Comment (jremmert-phytec-iot) [inline drivers/kw2xrf/kw2xrf_tx.c]: At this point it might be more efficient to call the write_fifo-function only once and write the complete packet at once via SPI.

Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_tx.c]: I totally agree, as `hdr` is a pointer to the same array as `pkt` is, it would just be easier to read.

Comment (jremmert-phytec-iot) [inline drivers/kw2xrf/kw2xrf_tx.c]: Agree, will use pkt at this point.

**Comment by jremmert-phytec-iot** [comment]
Tested this driver by applying #2265 (Kinetis_common) and #2059 (Phytec PBA-D-01).

I had to add a line to /tests/netdev/main.c to test it successfully (see below). Function "init_long_address" sets address to long-type. But the variable "dev_address_len", used for verification in the test-app is not updated to the long-address-type. May this be a bug, or am I misunderstanding that?

```c
static int check_long_address(void) 
     switch (res) {
         case 0:
+ +        dev_address_len = address_len;
             switch (address_len) {
                 case 1:
```

File: drivers/include/kw2xrf.h [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+#define MKW2XDRF_BROADCAST_ADDRESS (0xFFFF) /**< Broadcast address. */
+
+#define MKW2XDRF_DEFAULT_CHANNR      11     /**< Default radio channel. */
+
+/**
+ *  Structure to represent a kw2xrf packet.
+ */
+typedef struct __attribute__((packed))
+{
+    /* @{ */
```
Comment (haukepetersen) [inline drivers/include/kw2xrf.h]: is this compliant to doxygen?

File: drivers/kw2xrf/kw2xrf_spi.c (hunk 20 above)
Comment (thomaseichinger) [inline drivers/kw2xrf/kw2xrf_spi.c]: @jremmert-phytec-iot @jfischer-phytec-iot Although I don't agree with you argumentation I wont NACK it with using this separate SPI implementation. As long as you don't disable interrupts during SPI transactions below.

File: drivers/kw2xrf/include/kw2xrf_internal.h [hunk 28]
```diff
@@ -0,0 +1,291 @@
...
+ * @param[in] pan The desired pan ID.
+ *
+ * @return The set pan ID after calling.
+ */
+uint16_t kw2xrf_set_pan(uint16_t pan);
+
+/**
+ * @brief Gets the current IEEE long address of the KW2XRF.
+ *
+ * @return The current IEEE long address.
```
Comment (haukepetersen) [inline drivers/kw2xrf/include/kw2xrf_internal.h]: shouldn't this return the PAN id?

File: drivers/kw2xrf/kw2xrf.c [hunk 29]
```diff
@@ -0,0 +1,869 @@
...
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static uint16_t radio_pan;
+static uint8_t  radio_channel;
+static uint16_t radio_address;
+static uint64_t radio_address_long;
+
+#define MKW2XDRF_OUTPUT_POWER_MAX  8
```
Comment (haukepetersen) [inline drivers/kw2xrf/kw2xrf.c]: would be nice if you could document these defines.

File: drivers/kw2xrf/kw2xrf.c [hunk 30]
```diff
@@ -0,0 +1,869 @@
...
+#include "kernel_types.h"
+#include "netdev/802154.h"
+#include "netdev/base.h"
+#include "thread.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static uint16_t radio_pan;
```
Comment (haukepetersen) [inline drivers/kw2xrf/kw2xrf.c]: documentation would be nice here.

File: drivers/kw2xrf/kw2xrf.c [hunk 31]
```diff
@@ -0,0 +1,869 @@
...
+
+/* event type to signalize receive event */
+/* TODO: Event_Nr used from cc2420 driver */
+#define MKW2XDRF_NETDEV_EVENT_RX  (3452342219)
+
+/* default source address length for sending in number of byte */
+static size_t _default_src_addr_len = 2;
+
+/* PLL integer lookup table */
+static const uint8_t pll_int_lt[16] = {
```
Comment (haukepetersen) [inline drivers/kw2xrf/kw2xrf.c]: I would let the compiler put in the size of the array... Its just a source for error later on...

File: drivers/kw2xrf/kw2xrf.c [hunk 32]
```diff
@@ -0,0 +1,869 @@
...
+    11, 11, 11, 11,
+    11, 11, 12, 12,
+    12, 12, 12, 12,
+    13, 13, 13, 13
+};
+
+/* PLL frequency fractional lookup table */
+static const uint16_t pll_frac_lt[16] = {
```
Comment (haukepetersen) [inline drivers/kw2xrf/kw2xrf.c]: same here.

File: drivers/kw2xrf/kw2xrf.c [hunk 33]
```diff
@@ -0,0 +1,869 @@
...
+    10240, 20480, 30720, 40960,
+    51200, 61440, 6144, 16384,
+    26624, 36864, 47104, 57344,
+    2048, 12288, 22528, 32768
+};
+
+static const uint8_t pow_lt[44] = {
```
Comment (haukepetersen) [inline drivers/kw2xrf/kw2xrf.c]: and here (further occurrences are omitted)

File: drivers/kw2xrf/kw2xrf.c [hunk 34]
```diff
@@ -0,0 +1,869 @@
...
+        core_panic(0x42, "Could not start MKW2XD radio transceiver");
+    }
+
+    kw2xrf_set_tx_power(0);
+
+    /* Set radio pan id. */
+#ifdef MODULE_CONFIG
+    radio_pan = sysconfig.radio_pan_id;
+#else
+    radio_pan = 0x0001;
```
Comment (haukepetersen) [inline drivers/kw2xrf/kw2xrf.c]: I would suggest to use a define for the default pan ID

File: drivers/kw2xrf/kw2xrf_spi.c [hunk 35]
```diff
@@ -0,0 +1,277 @@
...
+    KW2XDRF_SPI->MCR = SPI_MCR_MSTR_MASK
+                       | SPI_MCR_PCSIS(1)
+                       | SPI_MCR_DOZE_MASK
+                       | SPI_MCR_CLR_TXF_MASK
+                       | SPI_MCR_CLR_RXF_MASK;
+
+    KW2XDRF_SPI->RSER = (uint32_t)0;
+
+    /* set up IRQ */
+    gpio_init_int(GPIO_KW2XDRF, GPIO_NOPULL, GPIO_FALLING, &kw2xrf_rx_irq, NULL);
```
Comment (haukepetersen) [inline drivers/kw2xrf/kw2xrf_spi.c]: looks strange to have this call inside the `spi_init`, I think I would move it to the general driver init() function

File: drivers/kw2xrf/kw2xrf_spi.c [hunk 36]
```diff
@@ -0,0 +1,277 @@
...
+    buf[i] = (uint8_t)KW2XDRF_SPI->POPR;
+    restoreIRQ(cpsr);
+}
+
+/* read / write packet buffer */
+
+radio_packet_length_t kw2xrf_write_fifo(uint8_t *data, radio_packet_length_t length)
+{
+    radio_packet_length_t i;
+    unsigned int cpsr = disableIRQ();
```
Comment (haukepetersen) [inline drivers/kw2xrf/kw2xrf_spi.c]: uff, disabling interrupts for quite a long time will definitely break RIOTs real-time properties, this should be avoided! Also it doesn't really hurt to jump into a interrupt routine while transferring data on the SPI bus when you are master (and therefore controlling the clock), so why disable it?
(same goes for all your SPI transfer functions!)

**Comment by jremmert-phytec-iot** [comment]
@haukepetersen Thanks for you comments! Indeed the IRQ-disabling is not neccessary and was ment for testing purposes. I addressed you comments accordingly and will updated the PR later this day.

There is an other question at the moment. So far, I only tested the driver with the /tests/netdev test. 
It would be important for us to have a working rpl-udp demo. In the Virtual meeting two weeks ago, we already discussed on this plan shortly. My conclusion was that it is possible with the PR "transceiver: netapi adapter #1969" to use the existing (old) network stack in combination with the new network device (this radio driver). 

1. Is this assumption correct in general? 
2. What steps would be neccessary to make the rpl_udp example runnable with the kw2x-rf radio driver? 
3. As is saw it, the default transceiver has to be defined in transceiver.h even through the transceiver interface is not used anymore? 

Sry about my confusion.

**Comment by haukepetersen** [comment]
@jremmert-phytec-iot The netdev (and netapi) are currently under heavy construction, and I honestly doubt we will get them stable before Embedded world... I personally do not know much about the netapi adapter, maybe @authmillenon can shed some light on this?

As a short-term solution towards the Embedded World I would probably suggest to implement the old transceiver interface for you radio. Although it will be thrown out at some point, you can run the current network stack (and the rpl_udp demo) with it. Also implementing this should only take a day or two...

**Comment by jremmert-phytec-iot** [comment]
Ok, im currently in the process of adaption. Your right, that might be the savest option to get a stable demo.

I think I missunderstood the netdev-adapter and the last days I couln't figure out how it works.

**Comment by jremmert-phytec-iot** [comment]
Adapted to Transceiver interface and rebased.

**Comment by jremmert-phytec-iot** [comment]
Closed due to refactoring and reimplementation in #2756.
